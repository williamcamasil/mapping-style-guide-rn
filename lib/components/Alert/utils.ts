import { ViewStyle } from 'react-native';

import { Icons } from '../../assets';
import { AppThemeType } from '../../theme/default';
import { ColorPaletesType } from '../../tokens';

export type AlertVariantType =
  'success' |
  'info' |
  'warning' |
  'error';

export const getVariantColorStyle = (variant?: AlertVariantType): ColorPaletesType => {
  switch (variant) {
    case 'warning':
      return 'feedbackAlert500';
    case 'error':
      return 'feedbackError500';
    case 'info':
      return 'neutralGray600';
    case 'success':
    default:
      return 'feedbackSuccess500';
  }
};

export const getDefaultIcon = (variant?: AlertVariantType) => {
  switch (variant) {
    case 'success':
      return Icons.Default.Success;
    case 'info':
      return Icons.Default.Information;
    case 'warning':
    case 'error':
    default:
      return Icons.Default.Warning;
  }
};

export const getAlertStyle = (
  theme: AppThemeType,
  variant?: AlertVariantType,
): ViewStyle => {

  switch (variant) {
    case 'warning':
      return {
        backgroundColor: theme.colors.feedbackAlert100,
      };
    case 'error':
      return {
        backgroundColor: theme.colors.feedbackError100,
      };
    case 'info':
      return {
        backgroundColor: theme.colors.neutralGray100,
      };
    case 'success':
    default:
      return {
        backgroundColor: theme.colors.feedbackSuccess100,
      };
  }
};
