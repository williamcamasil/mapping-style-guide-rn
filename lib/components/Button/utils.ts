import { ViewStyle } from 'react-native';

import { AppThemeType } from '../../theme/default';
import { ColorPaletesType } from '../../tokens';

export type ButtonVariantType =
  'containedPrimary' |
  'containedSecondary' |
  'outlinedPrimary' |
  'outlinedSecondary' |
  'text';

export type ButtonSizeType =
  'large' |
  'medium' |
  'small';

export const getVariantStyle = (
  theme: AppThemeType,
  variant?: ButtonVariantType,
  disabled?: boolean | null,
): ViewStyle => {
  switch (variant) {
    case 'containedSecondary':
      return {
        backgroundColor: disabled ? theme.colors.primary100 : theme.colors.primary200,
        borderWidth: theme.borders.width.none,
      };
    case 'outlinedPrimary':
      return {
        backgroundColor: 'transparent',
        borderWidth: theme.borders.width.thin,
        borderColor: disabled ? theme.colors.primary100 : theme.colors.primary600,
      };
    case 'outlinedSecondary':
      return {
        backgroundColor: 'transparent',
        borderWidth: theme.borders.width.thin,
        borderColor: disabled ? theme.colors.neutralGray100 : theme.colors.neutralGray300,
      };
    case 'text':
      return {
        backgroundColor: disabled ? theme.colors.neutralGray100 : 'transparent',
        borderWidth: theme.borders.width.none,
      };
    case 'containedPrimary':
    default:
      return {
        backgroundColor: disabled ? theme.colors.primary100 : theme.colors.primaryMain,
        borderWidth: theme.borders.width.none,
      };
  }
};

export const getTextColor = (
  variant?: ButtonVariantType,
  disabled?: boolean | null,
): ColorPaletesType => {
  switch (variant) {
    case 'containedSecondary':
      return disabled ? 'primary300' : 'primaryMain';
    case 'outlinedPrimary':
      return disabled ? 'primary300' : 'primaryMain';
    case 'outlinedSecondary':
      return disabled ? 'neutralGray200' : 'neutralGray600';
    case 'text':
      return disabled ? 'neutralGray400' : 'neutralGray500';
    case 'containedPrimary':
    default:
      return disabled ? 'primary300' : 'neutralWhite';
  }
};

export const getSizeStyle = (theme: AppThemeType, size?: ButtonSizeType, hasText?: boolean): ViewStyle => {
  switch (size) {
    case 'large':
      return {
        height: 64,
        minWidth: 64,
        paddingHorizontal: hasText ? theme.spacings.sXL : undefined,
      };
    case 'small':
      return {
        height: 30,
        minWidth: 30,
        paddingHorizontal: hasText ? theme.spacings.sXS : undefined,
      };
    case 'medium':
    default:
      return {
        height: 48,
        minWidth: 48,
        paddingHorizontal: hasText ? theme.spacings.sMedium : undefined,
      };
  }
};

export const getStylesToBlock = (block: boolean | undefined): ViewStyle | null => {
  if (block) return null;

  return {
    flexDirection: 'row',
  };
};
