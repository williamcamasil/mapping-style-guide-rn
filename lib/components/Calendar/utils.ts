import { dateToString } from 'mapping-context-rn';
import type { MarkedDates, Theme } from 'react-native-calendars/src/types';

import { AppThemeType } from '../../theme';

type ExtendedTheme = Theme & {
  'stylesheet.calendar.header': object;
  'stylesheet.day.basic': object;
};

/**
 * Define o tipo de data para o formato yyyy-MM-dd
 */
export type StringDateType = `${number}${number}${number}${number}-${number}${number}-${number}${number}`;

export const DATE_FORMAT = 'yyyy-MM-dd';

export const DAY_SIZE = 42;

export const getDefaultValue = (value?: StringDateType | null) => (
  value ?? dateToString(new Date(), DATE_FORMAT) as StringDateType
);

export const getDisabledDates = (disabledDates?: StringDateType[]) => {
  if (!disabledDates?.length) return {};
  const dates: MarkedDates = {};
  disabledDates.forEach((date: StringDateType) => {
    dates[date] = {
      disabled: true,
      disableTouchEvent: true,
    };
  });
  return dates;
};

export const getCommonTheme = (theme: AppThemeType): ExtendedTheme => ({
  selectedDayBackgroundColor: theme.colors.primaryMain,
  todayTextColor: theme.colors.primaryMain,
  dayTextColor: theme.colors.neutralGray600,
  textDayFontFamily: theme.typography.weights.bold.fontFamily,
  textMonthFontFamily: theme.typography.weights.bold.fontFamily,
  textDayHeaderFontFamily: theme.typography.weights.bold.fontFamily,
  textDayFontWeight: theme.typography.weights.bold.fontWeight,
  textMonthFontWeight: theme.typography.weights.bold.fontWeight,
  textDayHeaderFontWeight: theme.typography.weights.bold.fontWeight,
  textDayFontSize: theme.typography.sizes.small,
  textMonthFontSize: theme.typography.sizes.small,
  textDayHeaderFontSize: theme.typography.sizes.small,
  'stylesheet.calendar.header': {
    dayHeader: {
      marginTop: theme.spacings.sXS,
      marginBottom: theme.spacings.sXS,
      textAlign: 'center',
      fontSize: theme.typography.sizes.small,
      color: theme.colors.primaryMain,
      fontFamily: theme.typography.weights.bold.fontFamily,
      fontWeight: theme.typography.weights.bold.fontWeight,
      flex: 1,
    },
    week: {
      flexDirection: 'row',
      justifyContent: 'space-between',
    },
  },
  'stylesheet.day.basic': {
    base: {
      height: DAY_SIZE,
      width: DAY_SIZE,
      borderBottomLeftRadius: theme.borders.radius.XL,
      borderBottomRightRadius: theme.borders.radius.XL,
      borderTopLeftRadius: theme.borders.radius.XL,
      borderTopRightRadius: theme.borders.radius.XL,
      alignItems: 'center',
      justifyContent: 'center',
    },
  },
});
