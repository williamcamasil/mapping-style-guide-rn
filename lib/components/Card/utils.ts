import { ColorValue, ViewStyle } from 'react-native';

import { AppThemeType } from '../../theme/default';

export type CardSizeType =
  'small' |
  'big';

export type CardVariantType =
  'default' |
  'highlighted50' |
  'highlighted100';

export const getCardSizeStyle = (theme: AppThemeType, size?: CardSizeType): ViewStyle => {
  switch (size) {
    case 'small':
      return {
        width: 135,
        height: 120,
        padding: theme.spacings.sXS,
      };
    case 'big':
    default:
      return {
        minHeight: 128,
        padding: theme.spacings.sSmall,
      };
  }
};

export const getCardShadowStyle = (theme: AppThemeType, showShadow?: boolean): ViewStyle | null => {
  if (showShadow) {
    return theme.shadows.level4;
  }

  return {
    borderWidth: theme.borders.width.thin,
    borderColor: theme.colors.neutralGray200,
  };
};

export const getCardVariantStyle = (theme: AppThemeType, cardType?: CardVariantType): ViewStyle | null => {
  switch (cardType) {
    case 'highlighted50':
      return {
        borderWidth: theme.borders.width.thin,
        borderColor: theme.colors.primary700,
      };
    case 'highlighted100':
      return {
        borderWidth: theme.borders.width.thin,
        borderColor: theme.colors.primary600,
        backgroundColor: theme.colors.primaryMain,
      };
    case 'default':
    default:
      return null;
  }
};

export const getCardCustomBorderStyle = (theme: AppThemeType, borderColor?: ColorValue): ViewStyle | null => {
  if (!borderColor) return null;

  return {
    borderWidth: theme.borders.width.thin,
    borderColor,
  };
};

export const getArrowColor = (theme: AppThemeType, variant?: CardVariantType, arrowColor?: ColorValue): ColorValue => {
  if (arrowColor) return arrowColor;

  if (variant === 'highlighted100') return theme.colors.neutralWhite;

  return theme.colors.neutralGray500;
};
