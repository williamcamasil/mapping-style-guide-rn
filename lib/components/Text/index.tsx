import React from 'react';
import {
  StyleProp, StyleSheet, Text as RNText, TextProps as RNTextProp, TextStyle,
} from 'react-native';

import { useTextStyles } from '../../hooks';
import { withTheme } from '../../theme';
import { AppThemeType } from '../../theme/default';
import {
  ColorPaletesType, FontWeightType, LineHeightType, Typography,
} from '../../tokens';
import {
  getColorStyle, getVariantStyle, getWeightStyle, TextVariantType,
} from './utils';

export type { TextVariantType } from './utils';

export type TextPropTypes = RNTextProp & {
  theme: AppThemeType;
  style?: StyleProp<TextStyle>;
  color?: ColorPaletesType;
  variant?: TextVariantType;
  weight?: FontWeightType;
  lineHeight?: LineHeightType;
};

const styles = StyleSheet.create({
  text: {
    letterSpacing: Typography.letterSpaces.default,
    flexShrink: 1,
  },
});

const Text: React.FC<TextPropTypes> = ({
  children,
  color,
  theme,
  style,
  variant = 'body',
  weight,
  lineHeight,
  allowFontScaling = false,
  ...others
}) => {

  const textStyles = useTextStyles(() => [
    styles.text,
    getColorStyle(theme, color ?? 'neutralGray600'),
    getVariantStyle(theme, variant, lineHeight),
    getWeightStyle(theme, weight),
    style,
  ], [color, lineHeight, style, theme, variant, weight]);

  return (
    <RNText
      {...others}
      allowFontScaling={allowFontScaling}
      style={textStyles}
    >
      {children}
    </RNText>
  );
};

export default withTheme(Text);
