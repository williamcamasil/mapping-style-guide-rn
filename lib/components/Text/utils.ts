import { TextStyle } from 'react-native';

import { AppThemeType } from '../../theme/default';
import { ColorPaletesType, FontWeightType, LineHeightType } from '../../tokens';
import { calcLineHeight } from '../../utils/typography';

export type TextVariantType =
  // heading
  'headingDisplay' |
  'headingXLarge' |
  'headingLarge' |
  'headingMedium' |
  'headingSmall' |
  // paragraph regular
  'subtitle' |
  'body' |
  'bodySmall' |
  'caption' |
  // small
  'small' |
  'smallCaps' |
  // one line
  'oneLineHuge' |
  'oneLineLarge' |
  'oneLineMedium' |
  'oneLineSmall' |
  'oneLineCaption' |
  'oneLineSemibold' |
  'oneLineBoldHuge' |
  'oneLineBoldExtraLarge' |
  'oneLineBoldLarge' |
  'oneLineBoldMedium' |
  'oneLineBoldSmall';

const getHeadingStyles = (theme: AppThemeType, lineHeight: number | null) => ({
  headingDisplay: {
    ...theme.typography.weights.bold,
    fontSize: theme.typography.sizes.huge,
    lineHeight: calcLineHeight(theme.typography.sizes.huge, lineHeight ?? theme.typography.lineHeights.medium),
  },
  headingXLarge: {
    ...theme.typography.weights.bold,
    fontSize: theme.typography.sizes.XXXL,
    lineHeight: calcLineHeight(theme.typography.sizes.XXXL, lineHeight ?? theme.typography.lineHeights.medium),
  },
  headingLarge: {
    ...theme.typography.weights.bold,
    fontSize: theme.typography.sizes.XXL,
    lineHeight: calcLineHeight(theme.typography.sizes.XXL, lineHeight ?? theme.typography.lineHeights.large),
  },
  headingMedium: {
    ...theme.typography.weights.bold,
    fontSize: theme.typography.sizes.XL,
    lineHeight: calcLineHeight(theme.typography.sizes.XL, lineHeight ?? theme.typography.lineHeights.medium),
  },
  headingSmall: {
    ...theme.typography.weights.bold,
    fontSize: theme.typography.sizes.large,
    lineHeight: calcLineHeight(theme.typography.sizes.large, lineHeight ?? theme.typography.lineHeights.medium),
  },
});

const getSmallStyles = (theme: AppThemeType, lineHeight: number | null) => ({
  small: {
    ...theme.typography.weights.bold,
    fontSize: theme.typography.sizes.XS,
    lineHeight: calcLineHeight(14, lineHeight ?? theme.typography.lineHeights.medium),
  },
  smallCaps: {
    ...theme.typography.weights.bold,
    fontSize: theme.typography.sizes.XXXS,
    lineHeight: calcLineHeight(theme.typography.sizes.XXXS, lineHeight ?? theme.typography.lineHeights.medium),
    textTransform: 'uppercase' as TextStyle['textTransform'],
  },
});

const getOneLineStyles = (theme: AppThemeType, lineHeight: number | null) => ({
  oneLineHuge: {
    ...theme.typography.weights.regular,
    fontSize: theme.typography.sizes.XL,
    lineHeight: calcLineHeight(theme.typography.sizes.XL, lineHeight ?? theme.typography.lineHeights.small),
  },
  oneLineLarge: {
    ...theme.typography.weights.regular,
    fontSize: theme.typography.sizes.large,
    lineHeight: calcLineHeight(theme.typography.sizes.large, lineHeight ?? theme.typography.lineHeights.small),
  },
  oneLineMedium: {
    ...theme.typography.weights.regular,
    fontSize: theme.typography.sizes.small,
    lineHeight: calcLineHeight(theme.typography.sizes.small, lineHeight ?? theme.typography.lineHeights.small),
  },
  oneLineSmall: {
    ...theme.typography.weights.regular,
    fontSize: theme.typography.sizes.XS,
    lineHeight: calcLineHeight(theme.typography.sizes.XS, lineHeight ?? theme.typography.lineHeights.small),
  },
  oneLineCaption: {
    ...theme.typography.weights.regular,
    fontSize: theme.typography.sizes.XXS,
    lineHeight: calcLineHeight(theme.typography.sizes.XXS, lineHeight ?? theme.typography.lineHeights.small),
  },
  oneLineSemibold: {
    ...theme.typography.weights.semiBold,
    fontSize: theme.typography.sizes.medium,
    lineHeight: calcLineHeight(theme.typography.sizes.medium, lineHeight ?? theme.typography.lineHeights.small),
  },
  oneLineBoldHuge: {
    ...theme.typography.weights.bold,
    fontSize: theme.typography.sizes.XL,
    lineHeight: calcLineHeight(theme.typography.sizes.XL, lineHeight ?? theme.typography.lineHeights.small),
  },
  oneLineBoldExtraLarge: {
    ...theme.typography.weights.bold,
    fontSize: theme.typography.sizes.large,
    lineHeight: calcLineHeight(theme.typography.sizes.large, lineHeight ?? theme.typography.lineHeights.small),
  },
  oneLineBoldLarge: {
    ...theme.typography.weights.bold,
    fontSize: theme.typography.sizes.medium,
    lineHeight: calcLineHeight(theme.typography.sizes.medium, lineHeight ?? theme.typography.lineHeights.small),
  },
  oneLineBoldMedium: {
    ...theme.typography.weights.bold,
    fontSize: theme.typography.sizes.small,
    lineHeight: calcLineHeight(theme.typography.sizes.small, lineHeight ?? theme.typography.lineHeights.small),
  },
  oneLineBoldSmall: {
    ...theme.typography.weights.bold,
    fontSize: theme.typography.sizes.XS,
    lineHeight: calcLineHeight(theme.typography.sizes.XS, lineHeight ?? theme.typography.lineHeights.small),
  },
});

const getParagraphStyles = (theme: AppThemeType, lineHeight: number | null) => ({
  subtitle: {
    ...theme.typography.weights.regular,
    fontSize: theme.typography.sizes.medium,
    lineHeight: calcLineHeight(theme.typography.sizes.medium, lineHeight ?? theme.typography.lineHeights.large),
  },
  bodySmall: {
    ...theme.typography.weights.regular,
    fontSize: theme.typography.sizes.XS,
    lineHeight: calcLineHeight(theme.typography.sizes.XS, lineHeight ?? theme.typography.lineHeights.large),
  },
  caption: {
    ...theme.typography.weights.regular,
    fontSize: theme.typography.sizes.XXS,
    lineHeight: calcLineHeight(theme.typography.sizes.XXS, lineHeight ?? theme.typography.lineHeights.medium),
  },
  body: {
    ...theme.typography.weights.regular,
    fontSize: theme.typography.sizes.small,
    lineHeight: calcLineHeight(theme.typography.sizes.small, lineHeight ?? theme.typography.lineHeights.large),
  },
});

export function getVariantStyle(
  theme: AppThemeType,
  variant: TextVariantType,
  lineHeightName?: LineHeightType,
): TextStyle {
  const lineHeight = lineHeightName ? theme.typography.lineHeights[lineHeightName] : null;

  const variantStyles: Record<TextVariantType, TextStyle> = {
    // heading
    ...getHeadingStyles(theme, lineHeight),
    // small
    ...getSmallStyles(theme, lineHeight),
    // one line
    ...getOneLineStyles(theme, lineHeight),
    // paragraph
    ...getParagraphStyles(theme, lineHeight),
  };

  return variantStyles[variant];
}

export function getColorStyle(theme: AppThemeType, color: ColorPaletesType): TextStyle | null {
  return {
    color: theme.colors[color],
  };
}

export function getWeightStyle(theme: AppThemeType, weight?: FontWeightType): TextStyle | null {
  if (!weight) { return null; }

  return theme.typography.weights[weight];
}
