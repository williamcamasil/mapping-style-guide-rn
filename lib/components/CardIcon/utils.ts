import { ColorValue } from 'react-native';

import { AppThemeType } from '../../theme/default';
import { CardVariantType } from '../Card/utils';

export const getIconColor = (
  theme: AppThemeType,
  customColor?: ColorValue,
  variant?: CardVariantType,
): ColorValue | string => {
  if (customColor) return customColor;

  if (variant === 'highlighted100') {
    return theme.colors.neutralWhite;
  }

  return theme.colors.primaryMain;
};
