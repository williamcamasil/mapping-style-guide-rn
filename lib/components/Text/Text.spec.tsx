import React from 'react';
import { StyleSheet, Text as RNText } from 'react-native';

import { render } from '@testing-library/react-native';

import Text from '.';
import { DefaultTheme } from '../../theme';
import { ColorPaletesType, FontWeightType } from '../../tokens';
import { calcLineHeight } from '../../utils/typography';
import {
  getColorStyle, getVariantStyle, getWeightStyle, TextVariantType,
} from './utils';

const variants: TextVariantType[] = [
  // heading
  'headingDisplay',
  'headingXLarge',
  'headingLarge',
  'headingMedium',
  'headingSmall',
  // paragraph regular
  'subtitle',
  'body',
  'bodySmall',
  'caption',
  // small
  'small',
  'smallCaps',
  // one line
  'oneLineHuge',
  'oneLineLarge',
  'oneLineMedium',
  'oneLineSmall',
  'oneLineCaption',
  'oneLineSemibold',
  'oneLineBoldHuge',
  'oneLineBoldExtraLarge',
  'oneLineBoldLarge',
  'oneLineBoldMedium',
  'oneLineBoldSmall',
];

describe('Text', () => {
  it('renders correctly', () => {
    render(
      <Text>
        Hellow World
      </Text>,
    );
  });

  it('test children', () => {
    const testRenderer = render(
      <Text>
        Hellow World
      </Text>,
    );

    expect(testRenderer.UNSAFE_getByType(RNText).props.children).toBe('Hellow World');
  });
});

describe('Text styles', () => {
  it('color undefined', () => {
    const testRenderer = render(
      <Text color={undefined}>
        Hellow World
      </Text>,
    );

    const nativeInstance = testRenderer.UNSAFE_getByType(RNText);
    const nativeStyle = StyleSheet.flatten(nativeInstance.props.style);

    expect(nativeStyle).toHaveProperty('color', '#676D79');
  });

  it('test variant styles on React Native Text', () => {
    const renderTextWithVariant = (children: React.ReactNode, variant: TextVariantType) => {
      const testRenderer = render(
        <Text variant={variant} theme={DefaultTheme}>
          {children}
        </Text>,
      );

      const nativeInstance = testRenderer.UNSAFE_getByType(RNText);
      const nativeStyle = StyleSheet.flatten(nativeInstance.props.style);
      const variantStyles = getVariantStyle(DefaultTheme, variant);

      return {
        nativeStyle,
        variantStyles,
      };
    };

    variants.forEach(variant => {
      const { nativeStyle, variantStyles } = renderTextWithVariant(
        'Hellow World',
        variant,
      );

      expect(nativeStyle)
        .toEqual(expect.objectContaining(variantStyles));
    });
  });

  it('test color styles on React Native Text', () => {
    const renderTextWithColor = (children: React.ReactNode, color: ColorPaletesType) => {
      const testRenderer = render(
        <Text color={color} theme={DefaultTheme}>
          {children}
        </Text>,
      );

      const nativeInstance = testRenderer.UNSAFE_getByType(RNText);
      const nativeStyle = StyleSheet.flatten(nativeInstance.props.style);
      const colorStyles = getColorStyle(DefaultTheme, color);

      return {
        nativeStyle,
        colorStyles,
      };
    };

    const colors = Object.keys(DefaultTheme.colors) as ColorPaletesType[];

    colors.forEach(color => {
      const { nativeStyle, colorStyles } = renderTextWithColor(
        'Hellow World',
        color,
      );

      expect(nativeStyle)
        .toEqual(expect.objectContaining(colorStyles));
    });
  });

  it('test weight styles on React Native Text', () => {
    const renderTextWithWeight = (children: React.ReactNode, weight: FontWeightType) => {
      const testRenderer = render(
        <Text weight={weight} theme={DefaultTheme}>
          {children}
        </Text>,
      );

      const nativeInstance = testRenderer.UNSAFE_getByType(RNText);
      const nativeStyle = StyleSheet.flatten(nativeInstance.props.style);
      const weightStyles = getWeightStyle(DefaultTheme, weight);

      return {
        nativeStyle,
        weightStyles,
      };
    };

    const weights = Object.keys(DefaultTheme.typography.weights) as FontWeightType[];

    weights.forEach(weight => {
      const { nativeStyle, weightStyles } = renderTextWithWeight(
        'Hellow World',
        weight,
      );

      expect(nativeStyle)
        .toEqual(expect.objectContaining(weightStyles));
    });
  });

  it('test custom style', () => {
    const customStyles = {
      backgroundColor: 'red',
      fontSize: 50,
    };
    const testRenderer = render(
      <Text
        style={customStyles}
        variant="small"
      >
        Hellow World
      </Text>,
    );

    const nativeInstance = testRenderer.UNSAFE_getByType(RNText);
    const nativeStyle = StyleSheet.flatten(nativeInstance.props.style);

    expect(nativeStyle).toEqual(expect.objectContaining(customStyles));
  });

  it('test line height', () => {
    const testRenderer = render(
      <Text
        variant="small"
        lineHeight="large"
      >
        Hellow World
      </Text>,
    );

    const nativeInstance = testRenderer.UNSAFE_getByType(RNText);
    const nativeStyle = StyleSheet.flatten(nativeInstance.props.style);

    expect(nativeStyle.lineHeight).toEqual(20);
  });
});

describe('Text snapshot', () => {
  it.each(variants)('should match variant snapshot', variant => {
    const tree = render(
      <Text variant={variant}>
        {variant}
      </Text>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
  it('Should have the default variant as body', () => {
    const withoutVariant = render(
      <Text>
        Default variant
      </Text>,
    ).toJSON();
    const bodyVariant = render(
      <Text variant="body">
        Default variant
      </Text>,
    ).toJSON();

    expect(withoutVariant).toEqual(bodyVariant);
  });
});

describe('Text utils', () => {
  it('lineHeight', () => {
    expect(
      calcLineHeight(DefaultTheme.typography.sizes.small, DefaultTheme.typography.lineHeights.small),
    ).toBe(16);

    expect(
      calcLineHeight(DefaultTheme.typography.sizes.small, 0),
    ).toBeUndefined();
  });
});
