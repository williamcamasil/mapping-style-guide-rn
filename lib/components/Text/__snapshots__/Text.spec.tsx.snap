// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text snapshot should match variant snapshot 1`] = `
<Text
  allowFontScaling={false}
  style={
    {
      "color": "#676D79",
      "flexShrink": 1,
      "fontFamily": "Averta CY",
      "fontSize": 48,
      "fontWeight": "bold",
      "letterSpacing": -0.02,
      "lineHeight": 58,
    }
  }
>
  headingDisplay
</Text>
`;

exports[`Text snapshot should match variant snapshot 2`] = `
<Text
  allowFontScaling={false}
  style={
    {
      "color": "#676D79",
      "flexShrink": 1,
      "fontFamily": "Averta CY",
      "fontSize": 40,
      "fontWeight": "bold",
      "letterSpacing": -0.02,
      "lineHeight": 48,
    }
  }
>
  headingXLarge
</Text>
`;

exports[`Text snapshot should match variant snapshot 3`] = `
<Text
  allowFontScaling={false}
  style={
    {
      "color": "#676D79",
      "flexShrink": 1,
      "fontFamily": "Averta CY",
      "fontSize": 33,
      "fontWeight": "bold",
      "letterSpacing": -0.02,
      "lineHeight": 46,
    }
  }
>
  headingLarge
</Text>
`;

exports[`Text snapshot should match variant snapshot 4`] = `
<Text
  allowFontScaling={false}
  style={
    {
      "color": "#676D79",
      "flexShrink": 1,
      "fontFamily": "Averta CY",
      "fontSize": 27,
      "fontWeight": "bold",
      "letterSpacing": -0.02,
      "lineHeight": 32,
    }
  }
>
  headingMedium
</Text>
`;

exports[`Text snapshot should match variant snapshot 5`] = `
<Text
  allowFontScaling={false}
  style={
    {
      "color": "#676D79",
      "flexShrink": 1,
      "fontFamily": "Averta CY",
      "fontSize": 23,
      "fontWeight": "bold",
      "letterSpacing": -0.02,
      "lineHeight": 28,
    }
  }
>
  headingSmall
</Text>
`;

exports[`Text snapshot should match variant snapshot 6`] = `
<Text
  allowFontScaling={false}
  style={
    {
      "color": "#676D79",
      "flexShrink": 1,
      "fontFamily": "Averta CY",
      "fontSize": 19,
      "fontWeight": "normal",
      "letterSpacing": -0.02,
      "lineHeight": 27,
    }
  }
>
  subtitle
</Text>
`;

exports[`Text snapshot should match variant snapshot 7`] = `
<Text
  allowFontScaling={false}
  style={
    {
      "color": "#676D79",
      "flexShrink": 1,
      "fontFamily": "Averta CY",
      "fontSize": 16,
      "fontWeight": "normal",
      "letterSpacing": -0.02,
      "lineHeight": 22,
    }
  }
>
  body
</Text>
`;

exports[`Text snapshot should match variant snapshot 8`] = `
<Text
  allowFontScaling={false}
  style={
    {
      "color": "#676D79",
      "flexShrink": 1,
      "fontFamily": "Averta CY",
      "fontSize": 14,
      "fontWeight": "normal",
      "letterSpacing": -0.02,
      "lineHeight": 20,
    }
  }
>
  bodySmall
</Text>
`;

exports[`Text snapshot should match variant snapshot 9`] = `
<Text
  allowFontScaling={false}
  style={
    {
      "color": "#676D79",
      "flexShrink": 1,
      "fontFamily": "Averta CY",
      "fontSize": 13,
      "fontWeight": "normal",
      "letterSpacing": -0.02,
      "lineHeight": 16,
    }
  }
>
  caption
</Text>
`;

exports[`Text snapshot should match variant snapshot 10`] = `
<Text
  allowFontScaling={false}
  style={
    {
      "color": "#676D79",
      "flexShrink": 1,
      "fontFamily": "Averta CY",
      "fontSize": 14,
      "fontWeight": "bold",
      "letterSpacing": -0.02,
      "lineHeight": 17,
    }
  }
>
  small
</Text>
`;

exports[`Text snapshot should match variant snapshot 11`] = `
<Text
  allowFontScaling={false}
  style={
    {
      "color": "#676D79",
      "flexShrink": 1,
      "fontFamily": "Averta CY",
      "fontSize": 11,
      "fontWeight": "bold",
      "letterSpacing": -0.02,
      "lineHeight": 13,
      "textTransform": "uppercase",
    }
  }
>
  smallCaps
</Text>
`;

exports[`Text snapshot should match variant snapshot 12`] = `
<Text
  allowFontScaling={false}
  style={
    {
      "color": "#676D79",
      "flexShrink": 1,
      "fontFamily": "Averta CY",
      "fontSize": 27,
      "fontWeight": "normal",
      "letterSpacing": -0.02,
      "lineHeight": 27,
    }
  }
>
  oneLineHuge
</Text>
`;

exports[`Text snapshot should match variant snapshot 13`] = `
<Text
  allowFontScaling={false}
  style={
    {
      "color": "#676D79",
      "flexShrink": 1,
      "fontFamily": "Averta CY",
      "fontSize": 23,
      "fontWeight": "normal",
      "letterSpacing": -0.02,
      "lineHeight": 23,
    }
  }
>
  oneLineLarge
</Text>
`;

exports[`Text snapshot should match variant snapshot 14`] = `
<Text
  allowFontScaling={false}
  style={
    {
      "color": "#676D79",
      "flexShrink": 1,
      "fontFamily": "Averta CY",
      "fontSize": 16,
      "fontWeight": "normal",
      "letterSpacing": -0.02,
      "lineHeight": 16,
    }
  }
>
  oneLineMedium
</Text>
`;

exports[`Text snapshot should match variant snapshot 15`] = `
<Text
  allowFontScaling={false}
  style={
    {
      "color": "#676D79",
      "flexShrink": 1,
      "fontFamily": "Averta CY",
      "fontSize": 14,
      "fontWeight": "normal",
      "letterSpacing": -0.02,
      "lineHeight": 14,
    }
  }
>
  oneLineSmall
</Text>
`;

exports[`Text snapshot should match variant snapshot 16`] = `
<Text
  allowFontScaling={false}
  style={
    {
      "color": "#676D79",
      "flexShrink": 1,
      "fontFamily": "Averta CY",
      "fontSize": 13,
      "fontWeight": "normal",
      "letterSpacing": -0.02,
      "lineHeight": 13,
    }
  }
>
  oneLineCaption
</Text>
`;

exports[`Text snapshot should match variant snapshot 17`] = `
<Text
  allowFontScaling={false}
  style={
    {
      "color": "#676D79",
      "flexShrink": 1,
      "fontFamily": "Averta CY",
      "fontSize": 19,
      "fontWeight": "500",
      "letterSpacing": -0.02,
      "lineHeight": 19,
    }
  }
>
  oneLineSemibold
</Text>
`;

exports[`Text snapshot should match variant snapshot 18`] = `
<Text
  allowFontScaling={false}
  style={
    {
      "color": "#676D79",
      "flexShrink": 1,
      "fontFamily": "Averta CY",
      "fontSize": 27,
      "fontWeight": "bold",
      "letterSpacing": -0.02,
      "lineHeight": 27,
    }
  }
>
  oneLineBoldHuge
</Text>
`;

exports[`Text snapshot should match variant snapshot 19`] = `
<Text
  allowFontScaling={false}
  style={
    {
      "color": "#676D79",
      "flexShrink": 1,
      "fontFamily": "Averta CY",
      "fontSize": 23,
      "fontWeight": "bold",
      "letterSpacing": -0.02,
      "lineHeight": 23,
    }
  }
>
  oneLineBoldExtraLarge
</Text>
`;

exports[`Text snapshot should match variant snapshot 20`] = `
<Text
  allowFontScaling={false}
  style={
    {
      "color": "#676D79",
      "flexShrink": 1,
      "fontFamily": "Averta CY",
      "fontSize": 19,
      "fontWeight": "bold",
      "letterSpacing": -0.02,
      "lineHeight": 19,
    }
  }
>
  oneLineBoldLarge
</Text>
`;

exports[`Text snapshot should match variant snapshot 21`] = `
<Text
  allowFontScaling={false}
  style={
    {
      "color": "#676D79",
      "flexShrink": 1,
      "fontFamily": "Averta CY",
      "fontSize": 16,
      "fontWeight": "bold",
      "letterSpacing": -0.02,
      "lineHeight": 16,
    }
  }
>
  oneLineBoldMedium
</Text>
`;

exports[`Text snapshot should match variant snapshot 22`] = `
<Text
  allowFontScaling={false}
  style={
    {
      "color": "#676D79",
      "flexShrink": 1,
      "fontFamily": "Averta CY",
      "fontSize": 14,
      "fontWeight": "bold",
      "letterSpacing": -0.02,
      "lineHeight": 14,
    }
  }
>
  oneLineBoldSmall
</Text>
`;
